driver/example_test.go

```diff
@@ -149,3 +152,83 @@ func addAlbum(alb Album) (int64, error) {
 	}
 	return id, nil
 }
+
+func Example_customTime() {
+	// Get a database handle.
+	db, err := sql.Open("sqlite3", "file:/time.db?vfs=memdb")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer db.Close()
+
+	// Create a table to hold our data
+	_, err = db.Exec(`CREATE TABLE data (
+		id INTEGER PRIMARY KEY,
+		date_time TEXT
+		) STRICT`)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Initialise our custom time
+	c := CustomTime{time.Date(
+		2009, 11, 17, 20, 34, 58, 651387237, time.UTC)}
+
+	// Store our custom time in the DB
+	_, err = db.Exec(`INSERT INTO data (date_time) VALUES(?)`, c)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var strTime string
+	// Retrieve it as a string, the result of Value()
+	err = db.QueryRow(`
+		SELECT date_time
+		FROM data
+		WHERE id = last_insert_rowid()
+	`).Scan(&strTime)
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println(strTime)
+
+	var resTime CustomTime
+	// Retrieve it as our custom time type, going through Scan()
+	err = db.QueryRow(`
+		SELECT date_time
+		FROM data
+		WHERE id = last_insert_rowid()
+	`).Scan(&resTime)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	fmt.Println(resTime)
+	// Output:
+	// 2009-11-17T20:34:58.651387237Z
+	// 2009-11-17 20:34:58.651387237 +0000 UTC
+}
+
+type CustomTime struct{ time.Time }
+
+func (c CustomTime) Value() (driver.Value, error) {
+	return c.UTC().Format(time.RFC3339Nano), nil
+}
+
+func (c *CustomTime) Scan(value any) error {
+	switch v := value.(type) {
+	case nil:
+		*c = CustomTime{time.Time{}}
+	case time.Time:
+		*c = CustomTime{v}
+	case string:
+		t, err := time.Parse(time.RFC3339Nano, v)
```

Comment:
This might be a somewhat poor example since if it parsed correct as such `maybeTime` would've handled it and we would've gotten a `time.Time` instead.

I'm not sure how to modify the example to be more compelling, of `Exec()` writing a different text representation directly into the field and then decoding it with a different format here.

Reply:
A more compelling example might be using a format like `2006-01-02T15:04:05.000Z07:00`.

This is automatically revived most of the time, but sometimes isn't because of trailing zeros.

Reply:
I'm not entirely sure I follow. That string is valid for RFC3339Nano if I'm not mistaken. Would `maybeTime` not recover that?

Reply:
No, it wouldn't.

It's valid for `RFC3339Nano`, yes, but if you convert it to a time **and then** try to convert back to a string, you don't get the original string: the millisecond zeros are trimmed.

This would make it impossible to recover that string unmodified from the database, which goes against "don't destroy data." Any automatic default smartness on my end must be non-destructive.

The only exception with default settings is if you explicitly mark your column as a timestamp.

Reply:
Oh, right. It doesn't roundtrip cleanly. I'll update the example.

Reply:
It doesn't parse at all in that case:

```
t, err := time.Parse(time.RFC3339, "2006-01-02T15:04:05.000Z07:00")
```

That results in `parsing time "2006-01-02T15:04:05.000Z07:00": extra text: "07:00"`. Same if you use the RFC3339Nano format.

So there you always get a string, since RFC3339 expects an absence of nanoseconds and RFC3339Nano expects 9 digits.

If someone wrote that value directly instead of binding the custom time struct and thus going through driver.Valuer you'd hit this issue. You get a string, and then you need a custom time.Layout in Scan() when hitting a string.

Reply:
I meant `"2006-01-02T15:04:05.000Z07:00"` as a format, so:
```go
t, err := time.Parse("2006-01-02T15:04:05.000Z07:00", v)
```

If you use a format with trailing zeros, like `"2006-01-02T15:04:05.000Z07:00"`, 90% of the time it works because you get values like `"2009-11-17T20:34:58.651Z"`, but once in a blue moon, you get `"2009-11-17T20:34:58.650Z"` and `maybeTime` doesn't revive the value.

Also, using a fixed number of decimals (3 or 6) is often recommended because it makes strings lexically sortable.
Go is "in the wrong" here by preferring `RFC3339Nano`.

Reply:
Ah, now I see what you mean. I'll try and update the example and see if I can trigger that.